File: studyq-backend/models/attempt.py
from datetime import datetime, timezone
from models import db


class Attempt(db.Model):
    __tablename__ = "attempts"

    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey("users.id"), nullable=False)
    quiz_id = db.Column(db.Integer, db.ForeignKey("quizzes.id"), nullable=False)
    score = db.Column(db.Integer, nullable=False)
    total_questions = db.Column(db.Integer, nullable=False)
    attempted_at = db.Column(
        db.DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

    @property
    def accuracy(self) -> float:
        if self.total_questions == 0:
            return 0.0
        return round((self.score / self.total_questions) * 100, 1)

    def to_dict(self):
        return {
            "id": self.id,
            "user_id": self.user_id,
            "quiz_id": self.quiz_id,
            "score": self.score,
            "total_questions": self.total_questions,
            "accuracy": self.accuracy,
            "attempted_at": self.attempted_at.isoformat(),
        }

File: studyq-backend/models/Attempt.js
const mongoose = require('mongoose');

const attemptSchema = new mongoose.Schema({
  user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  quiz_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Quiz', required: true },
  score: { type: Number, required: true },
  total_questions: { type: Number, required: true },
  attempted_at: { type: Date, default: Date.now }
});

attemptSchema.virtual('accuracy').get(function() {
  if (this.total_questions === 0) return 0.0;
  return Number(((this.score / this.total_questions) * 100).toFixed(1));
});

attemptSchema.set('toObject', { virtuals: true });
attemptSchema.set('toJSON', { virtuals: true });

attemptSchema.methods.toDict = function() {
  return {
    id: this._id,
    user_id: this.user_id,
    quiz_id: this.quiz_id,
    score: this.score,
    total_questions: this.total_questions,
    accuracy: this.accuracy,
    attempted_at: this.attempted_at
  };
};

module.exports = mongoose.model('Attempt', attemptSchema);
